packages/flutter/lib/src/rendering/object.dart

```diff
@@ -1803,28 +1887,53 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
   @protected
   bool get alwaysNeedsCompositing => false;
 
-  OffsetLayer _layer;
   /// The compositing layer that this render object uses to repaint.
   ///
-  /// Call only when [isRepaintBoundary] is true and the render object has
-  /// already painted.
-  ///
-  /// To access the layer in debug code, even when it might be inappropriate to
-  /// access it (e.g. because it is dirty), consider [debugLayer].
-  OffsetLayer get layer {
-    assert(isRepaintBoundary, 'You can only access RenderObject.layer for render objects that are repaint boundaries.');
-    assert(!_needsPaint);
+  /// If this render object is not a repaint boundary, it is the responsibility
+  /// of the [paint] method to populate this field. If [needsCompositing] is
+  /// true, this field may be populated with the root-most layer used by the
+  /// render object implementation. When repainting, instead of creating a new
+  /// layer the render object may update the layer stored in this field for better
+  /// performance. It is also OK to leave this field as null and create a new
+  /// layer on every repaint, but without the performance benefit. If
+  /// [needsCompositing] is false, this field must be set to null either by
+  /// never populating this field, or by setting it to null when the value of
+  /// [needsCompositing] changes from true to false.
+  ///
+  /// If this render object is a repaint boundary, the framework automatically
+  /// creates an [OffsetLayer] and populates this field prior to calling the
+  /// [paint] method. The [paint] method must not replace the value of this
+  /// field.
+  ContainerLayer get layer {
```

Comment:
this should be protected. Nobody outside this object should be reading it as far as I can tell.

Reply:
There are 41 unprotected usages. Most are in tests (which can be fixed by making it `@protected @visibleForTesting`). However, there are 18 usages in `package:flutter_test`, `widget_inspector.dart`, and `rendering/binding.dart`, that need this to remain public.

Reply:
How many of these are for debug purposes (for which we could use a `debugLayer`) vs production?

Reply:
I was able to convert all but 5 to `debugLayer`. Here's the production code that uses `layer`:

rendering/binding.dart:

https://github.com/flutter/flutter/blob/b11c5acaa6200a8d662705a312a2e5fb7714fe2e/packages/flutter/lib/src/rendering/binding.dart#L249

multiple places in widget_inspector.dart (although I'm not sure if widget inspector is considered debug code):

https://github.com/flutter/flutter/blob/b11c5acaa6200a8d662705a312a2e5fb7714fe2e/packages/flutter/lib/src/widgets/widget_inspector.dart#L559

Reply:
Looking the surrounding code in `widget_inspector.dart` it's all debug code, so I was able to convert it to `debugLayer` as well. So, the only remaining production usage of `layer` is the `MouseTracker` in `rendering/binding.dart`.

Reply:
Let's move that closure into a method on RenderView, and then pass that method in rather than using a closure. That'll be generally cleaner and will allow us to make this protected.